odk1-src/form-logic.rst

```diff
@@ -0,0 +1,314 @@
+  | Tip (18%): $${tip_18}
+  | Total: $${tip_18_total}",
+
+
+
+
+.. _requiring_responses:
+
+Requiring responses
+=====================
+
+By default,
+users are able to skip questions in a form.
+To make a question required,
+put :tc:`yes` in the :th:`required` column.
+
+You can optionally include a :th:`required_message`
+which will be displayed to a user who tries to advance the form
+without answering the question.
+
+.. image
+
+.. rubric:: XLSForm
+
+.. csv-table:: survey
+  :header: type, name, label, required, required_message
+  
+  text, name, What is your name?, yes, Please answer the question.
+
+  
+  
+.. _relevants:
+
+Conditionally showing questions
+=================================
+
+The :th:`relevant` column can be used to
+show or hide
+questions and groups of questions
+based on previous responses.
+
+If the expression in the :th:`relevant` column
+evaluates to ``True``, 
+the question or group is shown.
+If ``False``, 
+the question is skipped.
+
+Often, :ref:`comparison operators` are used in relevance expressions. For example:
+
+:tc:`${age} <= 5`
+  True if :tc:`age` is five or less.
+  
+:tc:`${has_children} = 'yes'`
+  True if the answer to :tc:`has_children` was ``yes``.
+  
+Relevance expressions can also use `XPath functions`_.
+For example:
+
+.. _XPath functions: https://opendatakit.github.io/xforms-spec/#xpath-functions
+
+:tc:`selected(${allergies}, 'peanut')`
+  True if ``peanut`` was selected
+  in the :ref:`multi-select-widget` named :tc:`allergies`.
+  
+:tc:`contains(${haystack}, 'needle')`
+  True if the exact string ``needle`` 
+  is contained anywhere inside the response to :tc:`haystack`.
+  
+:tc:`count-selected(${toppings}) > 5`
+  True if more than five options were selected
+  in the :ref:`multi-select-widget` named :tc:`toppings`.
+
+
+.. _constraints:
+
+Validating and restricting responses
+=========================================
+
+To validate or restrict response values,
+use the :th:`constraint` column.
+The :th:`constraint` expression will be evaluated
+when the user advances to the next screen.
+If the expression evaluates to ``True``,
+the form advances as usual.
+If ``False``, 
+the form does not advance
+and the :th:`constraint_message` is displayed.
+
+The entered value of the response is represented in the expression
+with a single dot (``.``).
+
+Like :ref:`relevance expressions <relevants>`,
+constraint expressions often use comparison operators 
+and regular expressions.
+
+:tc:`. >= 18`
+  True if response is greater than or equal to 18.
+
+:tc:`. < 20 and . > 200`
+  True if the response is between 20 and 200.
+  
+:tc:`regex(.,'\p{L}+')`
+  True if the response only contains letters, without spaces, separators, or numbers.
+  
+:tc:`not(contains(., 'prohibited'))`
+  True if the substring ``prohibited`` does not appear in the response.
+
+  
+.. seealso:: :doc:`form-regex`
+
+
+.. _repeats:
+
+Repeating questions and groups of questions
```

Comment:
A separate page is available for repeats, how about moving this content there. Also, this content is a duplicate of http://xlsform.org/#repeats. How about redirecting the user to it and showing unique content in these docs?

Reply:
I think it's fine to link to xlsform.org for another example, but repeats are such a frequently asked about item, I think we should keep our own content.

Reply:
Maybe the /form-repeats should be retitled something like /form-repeat-recipes or something along those lines? Because right now we have some good repeat content in /form-logic already.

Reply:
That's literally what I called it at first. Then I thought it was basically an expansion of the repeat section of form-logic, and changed it (I also wasn't sure if "Recipes" was too informal or wouldn't translate well)

Reply:
my intention was to cover the basics of everything in form logic and then add additional pages for specific topics as needed,

Reply:
That seems like a reasonable approach! Because we could have form topics on repeats, media, styling, etc.